### DISCUSSION TITLE
Restart server on changes in non-reloadable files

### DISCUSSION THREAD

**Discussion by scarf005** [lumeland/lume#579] [Ideas]
## The Problem

When using live-reload unfriendly components like ESM modules (`.ts`) or JSX, you have to reload the server manually for every single changes. This is a lot of work.

adding `--watch` option also does not work:

`echo \"import 'lume/cli.ts'\" | deno run --watch=_includes/foo.ts --unstable -A -`

## Suggested Changes

allow users to configure server so the server would completely restart on file changes such as `.ts` or `.tsx`.

**Comment by oscarotero** [comment]
Yes, this is a limitation of Deno. It doesn't have any mecanism to invalidate an import and reload it, without restarting completely the script.

You can skip this limiation by using Lume components (files in _components folder) because they are not imported as esm (with `import name from "url"`). js/ts layers don't have this limitation neither because they are loaded using the same technique as components.
Does this fix your problem?

**Reply by scarf005** [reply]
they are live-reload based solution, however and can't be applied to some files (e.g `_config.ts`). it'd be nice to be able to  **completely restart** (== `ctrl+c` + `deno task serve`) server on changes that cannot be live reloaded.

would this be possible?

**Reply by oscarotero** [reply]
I have to dig into it.
Deno has some bugs with the way Lume is executed (using pipes like `echo [code to run] | deno run -A -`). And maybe this is why `--watch` doesn't work.
Can you try creating a file with the `import "lume/cli.ts"` code and run it directly (with `deno run --watch -A yourfile.ts`)?

**Reply by scarf005** [reply]
yeah, i was able to make lume watch files and fully restart via:

```json
{
	"tasks": {
		"serve": "deno run -A --watch=_components/,_includes main.ts -s"
	}
}
```

```sh
$ echo import "lume/cli.ts" > main.ts
$ deno task serve
```

**Reply by oscarotero** [reply]
Cool. I just remembered I was already filed this issue here: https://github.com/denoland/deno/issues/21069

**Reply by scarf005** [reply]
also with this way unstable-hmr works well. closing as workaround is working.

**Comment by ghost** [comment]
Nice! Here's a paraphrase for searchability, since it took me a long time to find this: when using code imported in `_config.ts`, such as your own plugins, live reload is not triggered. The above provides a workaround, using Deno's `--experimental-hmr` flag, where the entire server is restarted, which is at least an improvement over having to kill/restart the process by hand. More detail is provided in [this comment](https://github.com/lumeland/lume/issues/590#issuecomment-2031648988), which works for me on Deno 2.1.4 and Lume 2.4.3 (I presume this workaround is **not** made obsolete by `watcher.include` added in 2.3.1?)

**Reply by oscarotero** [reply]
`--experimental-hmr` is not a perfect solution, because it only works with imported modules, but Lume has to watch other files like `.md`, `.vto`, etc that are not loaded as modules.

For `_components`, the best way is to use the `comp` variable. More info: https://github.com/lumeland/lume/issues/647#issuecomment-2535377846

**Reply by ghost** [reply]
Hi @oscarotero, thanks for the response. Agreed it's not ideal, but from reading existing issues and discussions it seems like there's not a good way to do this with **plugins** yet due to Deno behaviour? Or am I missing something when it comes to authoring a local plugin, and expecting Lume to be able to watch the source? Seems to me you can't do [processing](https://lume.land/docs/core/processors/) in a [component](https://lume.land/docs/core/components/)?

e.g. `_config.ts`:

```typescript
import myPlugin from './plugins/my-plugin.ts';

site.use(myPlugin());
```

**Reply by oscarotero** [reply]
Lume reloads automatically after editing the _config.ts file. But cannot detect changes in modules imported by this file, like a local plugin. So in this case, yeah, the `--hmr` flag is the only way.
